# Student Performance Indicator

### LIfe Cycle of a Machine Learning Project

* Understanding the Problem Statement
* Data Collection 
* Data Validation
* Exploratory Data Analysis 
* Data Pre-Processing
* Model Training
* Choose Best Model

## 1.) Problem Statement

* This project understands how the student's performance (test scores) is affected by the other variables such as Gender, Ethnicity, Parental level of education, Lunch, and Test preparation course.

## 2.) Data Collection

* Data Source - https://www.kaggle.com/datasets/spscientist/students-performance-in-exams?datasetId=74977
* The data consists of * columns and 1000 rows. 

## 2.1) Import Data and Required Packages

In [1]:
#Import Dependencies 
import numpy as np
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt
%matplotlib inline
import warnings
warnings.filterwarnings('ignore')

## Extract CSV Data into Pandas Dataframe

In [2]:
#Create Student Performance Dataframe
sp_df = pd.read_csv('data/stud.csv')

#View Dataframe 
sp_df.head()

gender race_ethnicity parental_level_of_education         lunch  \
0  female        group B           bachelor's degree      standard   
1  female        group C                some college      standard   
2  female        group B             master's degree      standard   
3    male        group A          associate's degree  free/reduced   
4    male        group C                some college      standard   

  test_preparation_course  math_score  reading_score  writing_score  
0                    none          72             72             74  
1               completed          69             90             88  
2                    none          90             95             93  
3                    none          47             57             44  
4                    none          76             78             75

In [4]:
#View Dataframe Shape
sp_df.shape

(1000, 8)

## 2.2) Dataset Information 

* gender: Sex of Student -> (male/female)
* race_ethnicity: Ethnicity of the Student -> (Group A, B, C, D, E)
* parental_level_of_education: Parent's highest education completed -> (master's degree, bachelor's degree, associate's degree, some college, high school)
* lunch: Type of lunch -> (standard, free/reduced)
* test_preparation_course: supplemental material completion -> (completed, none)
* math_score
* reading_score
* writing_score  

## 3.) Data Validation

* Check missing values
* Check duplicates
* Check data types
* Check the number of unique values in each column
* Check statistics of the dataset
* Check various categories present in the columns 

## 3.1.) Check for Missing Values

In [5]:
#Count the amount of missing values
sp_df.isna().sum()

gender                         0
race_ethnicity                 0
parental_level_of_education    0
lunch                          0
test_preparation_course        0
math_score                     0
reading_score                  0
writing_score                  0
dtype: int64

* There are no missing values in the dataset

## 3.2.) Check for Duplicate Values

In [6]:
#Count the amount of duplicate values
sp_df.duplicated().sum()

0

* There are no duplicate values in the dataset

## 3.3.) Check Data Types

In [7]:
#Check for null values, and the data types in the dataset
sp_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 8 columns):
 #   Column                       Non-Null Count  Dtype 
---  ------                       --------------  ----- 
 0   gender                       1000 non-null   object
 1   race_ethnicity               1000 non-null   object
 2   parental_level_of_education  1000 non-null   object
 3   lunch                        1000 non-null   object
 4   test_preparation_course      1000 non-null   object
 5   math_score                   1000 non-null   int64 
 6   reading_score                1000 non-null   int64 
 7   writing_score                1000 non-null   int64 
dtypes: int64(3), object(5)
memory usage: 62.6+ KB


* There are no null values in the dataset

## 3.4.) Check the number of Unique Values in the Dataset's columns

In [8]:
#Check dataset for unique values
sp_df.nunique()

gender                          2
race_ethnicity                  5
parental_level_of_education     6
lunch                           2
test_preparation_course         2
math_score                     81
reading_score                  72
writing_score                  77
dtype: int64

## 3.5.) Check the Statistics of the Dataset

In [9]:
#Check basic stats of the dataset
sp_df.describe()

math_score  reading_score  writing_score
count  1000.00000    1000.000000    1000.000000
mean     66.08900      69.169000      68.054000
std      15.16308      14.600192      15.195657
min       0.00000      17.000000      10.000000
25%      57.00000      59.000000      57.750000
50%      66.00000      70.000000      69.000000
75%      77.00000      79.000000      79.000000
max     100.00000     100.000000     100.000000

## Insights

* From the description above of the numerical data, all means are very similar - between 65 and 68.05;
* All standard deviations are also very similar - between 14.6 and 15.19;
* While there is a minimum score of 0 for the math portion of the exam, the minimum socre for the writing portion is 10, and 17 for the reading portion of the exam

## 3.6) Exploring the Data